paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,497 @@
+
+  mkldnn::inner_product_forward::primitive_desc CreateFcPrimDesc(
+      const mkldnn::memory::desc& input_desc,
+      const mkldnn::memory::desc& weights_desc,
+      const mkldnn::memory::desc& dst_desc,
+      const mkldnn::primitive_attr& fc_attr) {
+    auto fc_desc = inner_product_forward::desc(prop_kind::forward, input_desc,
+                                               weights_desc, dst_desc);
+
+    return inner_product_forward::primitive_desc(fc_desc, fc_attr, engine_);
+  }
+
+  virtual mkldnn::memory CreateDstMemory(
+      const mkldnn::inner_product_forward::primitive_desc& fc_prim_desc,
+      const ExecutionContext& ctx, Tensor* output) {
+    auto dst_prim_desc = fc_prim_desc.dst_primitive_desc();
+    auto buffer_size = dst_prim_desc.get_size();
+
+    T* output_data = output->mutable_data<T>(
+        ctx.GetPlace(), ::paddle::memory::Allocator::kDefault, buffer_size);
+    output->set_format((memory::format)dst_prim_desc.desc().data.format);
+    return memory(dst_prim_desc, to_void_cast<T>(output_data));
+  }
+
+ protected:
+  const mkldnn::engine& engine_;
+  boost::optional<memory> bias_;
+  boost::optional<memory> input_;
+  boost::optional<memory> output_;
+  boost::optional<memory> weights_;
+  boost::optional<inner_product_forward> fc_;
+};  // namespace operators
+
+static std::string GetHash(const Tensor* input, const Tensor* weights,
+                           const std::string& suffix) {
+  auto dim2str = [](const DDim& operand_dims) {
+    std::string str = "";
+    for (int i = 0; i < operand_dims.size(); ++i) {
+      str += std::to_string(operand_dims[i]) + "-";
+    }
+    return str;
+  };
+
+  std::string hash = std::to_string((unsigned)input->format()) +
+                     std::to_string((unsigned)input->type()) +
+                     dim2str(weights->dims()) + suffix;
+
+  return hash;
+}
+
+template <typename T, typename K>
+class QuantMulPrimitiveFactory : public MulPrimitiveFactory<T, K> {
+ public:
+  using MulPrimitiveFactory<T, K>::MulPrimitiveFactory;
+
+  virtual inner_product_forward CreateMulPrimitive(
+      const Tensor* input, const Tensor* weights, Tensor* output,
+      const ExecutionContext& ctx) {
+    /* check data format and reorder if need */
+    int x_num_col_dims = ctx.Attr<int>("x_num_col_dims");
+    int y_num_col_dims = ctx.Attr<int>("y_num_col_dims");
+    auto scale_weight = ctx.Attr<std::vector<float>>("scale_y");
+
+    memory::data_type x_dt = framework::ToMKLDNNDataType(input->type());
+    memory::data_type y_dt = framework::ToMKLDNNDataType(weights->type());
+
+    auto x_matrix = this->UpdateDataFormat(input, x_num_col_dims, x_dt);
+    auto y_matrix = this->UpdateDataFormat(weights, y_num_col_dims, y_dt);
+
+    auto output_dim = output->dims();
+    if (output_dim.size() != 2) {
+      (*output).Resize({input->dims()[0], weights->dims()[1]});
+    }
+
+    if (this->fc_) {
+      this->UpdateDataPointers(ctx, output, input);
+      return *(this->fc_);
+    }
```

Comment:
Once again - this part should come after updating of weights, or otherwise the input data from y won't be transposed and quantized.

Reply:
since y to mul won't change, so it's unnecessary to update y

Reply:
done